File: WINGS_AI_Bootcamp/Python/sieve_of_eratosthenes_PrimeRange.py
def sieve_of_eratosthenes(n: int) -> list[int]:
    sieve = [True] * (n + 1)
    sieve[0], sieve[1] = False, False

    p = 2
    while p * p <= n:
        if sieve[p]:
            for i in range(p * p, n + 1, p):
                sieve[i] = False
        p += 1

    return [i for i, prime in enumerate(sieve) if prime]

# Example: primes up to 30
print(sieve_of_eratosthenes(30))

File: WINGS_AI_Bootcamp/Elixir/sieve_of_eratosthenes_PrimeRange.ex
defmodule Sieve do
  def sieve_of_eratosthenes(n) do
    if n < 2 do
      []
    else
      list = Enum.to_list(2..n)
      sieve(list, [])
      |> Enum.reverse()
    end
  end

  defp sieve([], primes), do: primes

  defp sieve([p | rest], primes) do
    filtered = Enum.filter(rest, fn x -> rem(x, p) != 0 end)
    sieve(filtered, [p | primes])
  end
end

IO.inspect(Sieve.sieve_of_eratosthenes(30))

